### PR TITLE
datetime functions

### PR DESCRIPTION
**Pull request by bheni**

### CODE REVIEW

File: enginetest/queries.go [hunk 1]
```diff
@@ -1724,14 +1724,6 @@ var QueryTests = []QueryTest{
 		`SELECT NOW() - NOW()`,
 		[]sql.Row{{int64(0)}},
 	},
-	{
-		`SELECT DATETIME() - NOW()`,
-		[]sql.Row{{int64(0)}},
-	},
-	{
-		`SELECT TIMESTAMP() - NOW()`,
```
Comment (bheni) [inline enginetest/queries.go:1732]: 0 argument versions not part of mysql

https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_timestamp

File: sql/expression/function/date.go [hunk 2]
```diff
@@ -160,7 +161,6 @@ func (d *DateSub) String() string {
 
 // TimestampConversion is a shorthand function for CONVERT(expr, TIMESTAMP)
 type TimestampConversion struct {
-	clock clock
```
Comment (bheni) [inline sql/expression/function/date.go:163]: remove no argument conversion as it's not supported in mysql

File: sql/expression/function/date_format.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+func (af AppendFuncWrapper) Append(bytes []byte, t time.Time) []byte {
+	s := af.fn(t)
+	return append(bytes, []byte(s)...)
+}
+
+var mysqlDateFormatSpec = strftime.NewSpecificationSet()
+var specifierToFunc = map[byte]func(time.Time) string {
+	'a': nil,
```
Comment (bheni) [inline sql/expression/function/date_format.go:169]: nil will remove strftime's default behavior for a specifier

File: sql/expression/function/user_test.go [hunk 4]
```diff
@@ -9,25 +9,26 @@ import (
 )
 
 func TestUser(t *testing.T) {
```
Comment (zachmu) [inline sql/expression/function/user_test.go:11]: This should probably go in some other file now that you've deleted the non-test file.

File: sql/expression/function/date_format.go [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+	}
+}
+
+func monthNum(t time.Time) string {
+	return strconv.FormatInt(int64(t.Month()), 10)
+}
+
+func dayWithSuffix(t time.Time) string {
+	suffix := "th"
+	day := int64(t.Day())
```
Comment (zachmu) [inline sql/expression/function/date_format.go:24]: uint8?

Comment (zachmu) [reply sql/expression/function/date_format.go:24]: Why this cast at all, really? Why not just leave it int?

File: sql/expression/function/date_format.go (hunk 3 above)
Comment (zachmu) [reply sql/expression/function/date_format.go:169]: Are these all of them? Not all letter cases are represented here

File: sql/expression/function/date_format.go [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+	if !ok {
+		return nil, ErrInvalidArgument.New("DATE_FORMAT", "format must be a string")
+	}
+
+
+	return strftime.Format(formatStr, t)
+}
+
+// Type implements the Expression interface.
+func (f *DateFormat) Type() sql.Type {
```
Comment (zachmu) [inline sql/expression/function/date_format.go:285]: Isn't this just String?

File: sql/expression/function/date_format.go [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+// Eval implements the Expression interface.
+func (f *DateFormat) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	left, err := f.Left.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if left != nil {
```
Comment (zachmu) [inline sql/expression/function/date_format.go:257]: This doesn't seem right

File: sql/expression/function/date_format.go [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+		return left, nil
+	}
+
```
Comment (zachmu) [inline sql/expression/function/date_format.go:260]: Need to handle left being nil. Should return NULL in that case

File: sql/expression/function/date_format.go [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+		if sql.IsNull(f.Right) {
+			return true
+		}
+		return f.Right.IsNullable()
+	}
+	return f.Left.IsNullable()
+}
+
+func (f *DateFormat) String() string {
+	return fmt.Sprintf("ifnull(%s, %s)", f.Left, f.Right)
```
Comment (zachmu) [inline sql/expression/function/date_format.go:307]: Fix

File: sql/expression/function/date_format_test.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+)
+
+func TestDateFormatting(t *testing.T) {
+	dt := time.Date(2020, 2,3,4,5,6,7000, time.UTC)
+	tests := []struct {
+		formatStr string
+		expected string
+		expectErr bool
+	} {
+		{"%a", "Mon", false}, // Abbreviated weekday name (Sun to Sat)
```
Comment (zachmu) [inline sql/expression/function/date_format_test.go:18]: No tests that expect an error?

Comment (zachmu) [reply sql/expression/function/date_format_test.go:18]: Also need tests for handling nil values for date and format

File: enginetest/queries.go [hunk 11]
```diff
@@ -1724,14 +1724,6 @@ var QueryTests = []QueryTest{
...
-		[]sql.Row{{int64(0)}},
-	},
 	{
 		`SELECT NOW() - (NOW() - INTERVAL 1 SECOND)`,
```
Comment (zachmu) [inline enginetest/queries.go:1728]: Should add a simple engine test for date_format

File: enginetest/queries.go (hunk 1 above)
Comment (bheni) [reply enginetest/queries.go:1732]: done

File: sql/expression/function/date.go [hunk 12]
```diff
@@ -266,31 +247,25 @@ func (t *DatetimeConversion) WithChildren(children ...sql.Expression) (sql.Expre
 }
 
 func NewDatetime(args ...sql.Expression) (sql.Expression, error) {
```
[the remark this answers is not in the document]
Comment (bheni) [reply sql/expression/function/date.go:249]: done

File: sql/expression/function/date.go [hunk 13]
```diff
@@ -266,31 +247,25 @@ func (t *DatetimeConversion) WithChildren(children ...sql.Expression) (sql.Expre
...
-	return &DatetimeConversion{defaultClock, args[0]}, nil
+
+	return &DatetimeConversion{args[0]}, nil
 }
 
 // UnixTimestamp converts the argument to the number of seconds since
-// 1970-01-01 00:00:00 UTC. With no argument, returns number of seconds since
-// unix epoch for the current time.
+// 1970-01-01 00:00:00 UTC.
 type UnixTimestamp struct {
```
[the remark this answers is not in the document]
Comment (bheni) [reply sql/expression/function/date.go:259]: fixed

File: sql/expression/function/date_format.go (hunk 5 above)
Comment (bheni) [reply sql/expression/function/date_format.go:24]: Has to be cast to pass to FormatInt.

File: sql/expression/function/date_format.go (hunk 7 above)
Comment (bheni) [reply sql/expression/function/date_format.go:257]: no, it's left

File: sql/expression/function/date_format.go (hunk 8 above)
Comment (bheni) [reply sql/expression/function/date_format.go:260]: fixed

File: sql/expression/function/date_format.go (hunk 6 above)
Comment (bheni) [reply sql/expression/function/date_format.go:285]: yes! Some copy pasta that I forgot to change.

File: sql/expression/function/date_format.go (hunk 9 above)
Comment (bheni) [reply sql/expression/function/date_format.go:307]: done

File: sql/expression/function/date_format_test.go (hunk 10 above)
Comment (bheni) [reply sql/expression/function/date_format_test.go:18]: done

File: sql/expression/function/time.go [hunk 14]
```diff
@@ -389,6 +394,97 @@ func (*YearWeek) WithChildren(children ...sql.Expression) (sql.Expression, error
...
+
+	mode := int64(0)
+	val, err := d.mode.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if val != nil {
+		if i64, err := sql.Int64.Convert(val); err == nil {
+			if mode, ok = i64.(int64); ok {
+				mode %= 8 // mode in [0, 7]
```
[the remark this answers is not in the document]
Comment (bheni) [reply sql/expression/function/time.go:454]: This is taken from the existing yearweek function.  I had assumed it matches mysql

[the remark this answers is not in the document]
Comment (bheni) [reply sql/expression/function/time.go:454]: verified that it also works this way in mysql

File: sql/expression/function/time.go [hunk 15]
```diff
@@ -501,45 +597,94 @@ var (
...
 // Children implements the sql.Expression interface.
-func (*Now) Children() []sql.Expression { return nil }
+func (n *Now) Children() []sql.Expression { return nil }
 
 // Eval implements the sql.Expression interface.
-func (n *Now) Eval(*sql.Context, sql.Row) (interface{}, error) {
-	return n.clock(), nil
+func (n *Now) Eval(ctx *sql.Context, _ sql.Row) (interface{}, error) {
+	t := ctx.QueryTime()
+	/*s, err := formatDate("%Y-%m-%d %H:%i:%s", t)
```
[the remark this answers is not in the document]
Comment (bheni) [reply sql/expression/function/time.go:658]: When we make this return a string in the future, this is the correct code.

File: sql/expression/function/time.go [hunk 16]
```diff
@@ -575,3 +720,91 @@ func (d *Date) WithChildren(children ...sql.Expression) (sql.Expression, error)
...
+func NewUnaryDatetimeFunc(name string, sqlType sql.Type, logic datetimeFuncLogic) sql.Function1 {
+	fn := func(e sql.Expression) sql.Expression {
+		return &UnaryDatetimeFunc{expression.UnaryExpression{Child: e}, name, sqlType, logic}
+	}
+
+	return sql.Function1{Name: name, Fn: fn}
+}
+
+// Eval implements the Expression interface.
+func (dtf *UnaryDatetimeFunc) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (bheni) [reply sql/expression/function/time.go:747]: I will not be bullied be internet slang it's a date time function.  It's dtf

File: sql/expression/function/date_test.go [hunk 17]
```diff
@@ -90,33 +90,21 @@ func TestDateSub(t *testing.T) {
 func TestUnixTimestamp(t *testing.T) {
 	require := require.New(t)
 
-	_, err := NewUnixTimestamp()
-	require.NoError(err)
-
-	_, err = NewUnixTimestamp(expression.NewLiteral("2018-05-02", sql.LongText))
+	_, err := NewUnixTimestamp(expression.NewLiteral("2018-05-02", sql.LongText))
```
[the remark this answers is not in the document]
Comment (bheni) [reply sql/expression/function/date_test.go:93]: done
